src/lib/HST/Frontend/FromGHC.hs

```diff
@@ -0,0 +1,382 @@
+-- Type Family Instances                                                     --
+-------------------------------------------------------------------------------
+
+-- | Type representing the AST data structure of @ghc-lib-parser@.
+--
+--   Instantiates the type families for source spans, literals and type
+--   expressions with the concrete types from @ghc-lib-parser@ or wrappers for
+--   these types. Also adds instances for 'S.EqAST' and 'S.ShowAST' to allow
+--   the usage of @==@ and @show@.
+data GHC
+type instance S.SrcSpanType GHC = GHC.SrcSpan
+type instance S.Literal GHC = LitWrapper
+type instance S.TypeExp GHC = TypeWrapper
+
+instance S.EqAST GHC
+instance S.ShowAST GHC
+
+-------------------------------------------------------------------------------
+-- Wrappers for @ghc-lib-parser@ Types                                       --
+-------------------------------------------------------------------------------
+
+-- | Wrapper for the two literal types (for regular and overloaded literals)
+--   used by @ghc-lib-parser@.
+data LitWrapper = Lit (GHC.HsLit GHC.GhcPs)
+                | OverLit (GHC.HsOverLit GHC.GhcPs)
+  deriving Eq
+
+instance Show LitWrapper where
+  show (Lit     l) = defaultPrintShow l
+  show (OverLit l) = defaultPrintShow l
+
+-- | Wrapper for the type for type expressions appearing in type signatures
+--   used by @ghc-lib-parser@.
+newtype TypeWrapper = SigType (GHC.LHsSigWcType GHC.GhcPs)
+
+instance Eq TypeWrapper where
+  SigType t1 == SigType t2 = defaultPrintEq t1 == defaultPrintEq t2
+
+instance Show TypeWrapper where
+  show (SigType t) = defaultPrintShow t
+
+-------------------------------------------------------------------------------
+-- Printing Functions for the @ghc-lib-parser@ AST                           --
+-------------------------------------------------------------------------------
+
+-- | Prints a component of the @ghc-lib-parser@ AST data structure without
+--   source span information and with line breaks.
+--
+--   This function is used for equality of @ghc-lib-parser@ AST components.
+defaultPrintEq :: Data a => a -> String
+defaultPrintEq d =
+  GHC.showSDoc defaultDynFlags (GHC.showAstData GHC.BlankSrcSpan d)
+
+-- | Prints a component of the @ghc-lib-parser@ AST data structure with source
+--   span information and without line breaks.
+--
+--   This function is used for showing @ghc-lib-parser@ AST components.
+defaultPrintShow :: Data a => a -> String
+defaultPrintShow d =
+  GHC.showSDocOneLine defaultDynFlags (GHC.showAstData GHC.NoBlankSrcSpan d)
+
+-- | Returns default variants of the dynamic flags used by @ghc-lib-parser@.
+--
+--   These dynamic flags are used for parsing Haskell code and printing AST
+--   components.
+defaultDynFlags :: GHC.DynFlags
+-- TODO These DynFlags should be the same as the ones used for parsing Haskell
+-- modules and should be defined in another module.
+defaultDynFlags = GHC.defaultDynFlags GHC.fakeSettings GHC.fakeLlvmConfig
```

Comment:
Since the type family instances start to become more complicated, I'd suggest moving them to their own modules (e.g. `HST.Frontend.Syntax.GHC` and `HST.Frontend.Syntax.HSE`)

Reply:
I would call these modules `HST.Frontend.GHC` and `HST.Frontend.HSE`, so without `Syntax`, because they are not a part of our syntax.
We could also create subdirectories `GHC` and `HSE` in the `Frontend` directory and call these modules `HST.Frontend.GHC.GHCConfig` and `HST.Frontend.GHC.HSEConfig` or similar.

Reply:
I considered the `type instance`s part of the syntax since the `type family` declarations are in the `Syntax` module, too. However, `GHC` and `HSE` subdirectories are fine as well. `From*` and `To*` could be moved to the subdirectories as well. I'd just avoid using `GHC` and `HSE` in module names twice, i.e., `HST.Frontend.GHC.Config` and `HST.Frontend.GHC.From` instead of `HST.Frontend.GHC.GHCConfig` and `HST.Frontend.GHC.FromGHC`.